### PR TITLE
Let icc also dump ICC profiles embedded in jpeg files

### PR DESCRIPTION
**Pull request by nico**
Adds a slightly rickety API to ImageDecoder for extracting ICC bytes from an image file.

Eventually the API should be fairly different. Eventually, all color data in the system should be tagged with some colorspace, so maybe Bitmap will just always contain a ref to a color space object or something.

---

Screenshot:

```
% time Build/lagom/icc ~/src/jpegfiles/DSC03136-exported.jpg
    preferred CMM type: 'Lino'
               version: 2.1.0
          device class: DisplayDevice
      data color space: RGB
      connection space: PCSXYZ
creation date and time: 1998-02-09 01:49:00
      primary platform: Microsoft
                 flags: 0x00000000
                        - not embedded in file
                        - can be used independently of embedded color data
   device manufacturer: 'IEC '
          device model: 'sRGB'
     device attributes: 0x0000000000000000
                        media is:
                        - reflective
                        - glossy
                        - of positive polarity
                        - colored
      rendering intent: Perceptual
        pcs illuminant: X = 0.964202, Y = 1, Z = 0.824905
               creator: 'HP  '
                    id: (not set)

tags:
copyrightTag ('cprt'): type 'text', offset 336, size 51
    text: "Copyright (c) 1998 Hewlett-Packard Company"
profileDescriptionTag ('desc'): type 'desc', offset 388, size 108
    ascii: "sRGB IEC61966-2.1"
    unicode: (not set)
    unicode language code: 0x0
    macintosh: "sRGB IEC61966-2.1"
mediaWhitePointTag ('wtpt'): type 'XYZ ', offset 496, size 20
    X = 0.950454, Y = 1, Z = 1.08905
Unknown tag ('bkpt'): type 'XYZ ', offset 516, size 20
    X = 0, Y = 0, Z = 0
redMatrixColumnTag ('rXYZ'): type 'XYZ ', offset 536, size 20
    X = 0.436065, Y = 0.222488, Z = 0.013916
greenMatrixColumnTag ('gXYZ'): type 'XYZ ', offset 556, size 20
    X = 0.385147, Y = 0.716873, Z = 0.097076
blueMatrixColumnTag ('bXYZ'): type 'XYZ ', offset 576, size 20
    X = 0.143066, Y = 0.060607, Z = 0.714096
deviceMfgDescTag ('dmnd'): type 'desc', offset 596, size 112
    ascii: "IEC http://www.iec.ch"
    unicode: (not set)
    unicode language code: 0x0
    macintosh: "IEC http://www.iec.ch"
deviceModelDescTag ('dmdd'): type 'desc', offset 708, size 136
    ascii: "IEC 61966-2.1 Default RGB colour space - sRGB"
    unicode: (not set)
    unicode language code: 0x0
    macintosh: "IEC 61966-2.1 Default RGB colour space - sRGB"
viewingCondDescTag ('vued'): type 'desc', offset 844, size 134
    ascii: "Reference Viewing Condition in IEC61966-2.1"
    unicode: (not set)
    unicode language code: 0x0
    macintosh: "Reference Viewing Condition in IEC61966-2.1"
viewingConditionsTag ('view'): type 'view', offset 980, size 36
luminanceTag ('lumi'): type 'XYZ ', offset 1016, size 20
    X = 76.036468, Y = 80, Z = 87.124618
measurementTag ('meas'): type 'meas', offset 1036, size 36
technologyTag ('tech'): type 'sig ', offset 1072, size 12
redTRCTag ('rTRC'): type 'curv', offset 1084, size 2060
    curve with 1024 points
greenTRCTag ('gTRC'): type 'curv', offset 1084, size 2060
    (see 'rTRC' above)
blueTRCTag ('bTRC'): type 'curv', offset 1084, size 2060
    (see 'rTRC' above)
```

### CODE REVIEW

File: Userland/Utilities/icc.cpp [hunk 1]
```diff
@@ -31,12 +32,24 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
-    static StringView icc_path;
-    args_parser.add_positional_argument(icc_path, "Path to ICC profile", "FILE");
+    static StringView path;
+    args_parser.add_positional_argument(path, "Path to ICC profile or to image containing ICC profile", "FILE");
     args_parser.parse(arguments);
 
-    auto icc_file = TRY(Core::MappedFile::map(icc_path));
-    auto profile = TRY(Gfx::ICC::Profile::try_load_from_externally_owned_memory(icc_file->bytes()));
+    auto file = TRY(Core::MappedFile::map(path));
+    auto icc_bytes = file->bytes();
```
Comment (linusg) [inline Userland/Utilities/icc.cpp:40]: Might be better to call this something like `file_bytes`, given we don't actually know what it is at this point, and then assign to a `ReadonlyBytes icc_bytes;` in the `if (decoder) ... else ...` branches.

**Review by linusg — COMMENTED** [review]
`Add a method _to_`

File: Userland/Libraries/LibGfx/BMPLoader.cpp [hunk 2]
```diff
@@ -1492,4 +1492,8 @@ ErrorOr<ImageFrameDescriptor> BMPImageDecoderPlugin::frame(size_t index)
     return ImageFrameDescriptor { m_context->bitmap, 0 };
 }
 
+ErrorOr<Optional<ReadonlyBytes>> BMPImageDecoderPlugin::icc_data()
+{
+    return OptionalNone {};
+}
```
[the remark this answers is not in the document]
Comment (nico) [reply Userland/Libraries/LibGfx/BMPLoader.cpp:1498]: clang-format actively _removes_ this newline:

```
 % git clang-format master
changed files:
    Userland/Libraries/LibGfx/BMPLoader.cpp
thakis@Nicos-MacBook-Pro serenity % git diff
diff --git a/Userland/Libraries/LibGfx/BMPLoader.cpp b/Userland/Libraries/LibGfx/BMPLoader.cpp
index 5cf8e8ca37..7606d24fa0 100644
--- a/Userland/Libraries/LibGfx/BMPLoader.cpp
+++ b/Userland/Libraries/LibGfx/BMPLoader.cpp
@@ -1496,5 +1496,4 @@ ErrorOr<Optional<ReadonlyBytes>> BMPImageDecoderPlugin::icc_data()
 {
     return OptionalNone {};
 }
-
 }
```

Still want me to add it? Seems like presubmit might fail then.

File: Userland/Utilities/icc.cpp (hunk 1 above)
Comment (nico) [reply Userland/Utilities/icc.cpp:40]: Done.

**Comment by nico** [comment]
Assigned to `icc_bytes` in the else, and fixed commit message. See comment above for newline before closing namespace `}`.

File: Userland/Libraries/LibGfx/BMPLoader.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (linusg) [reply Userland/Libraries/LibGfx/BMPLoader.cpp:1498]: it's there now, why would it do that 😐

https://github.com/SerenityOS/serenity/blob/6cfb057430df4a9f92bc29b670a0a72aeb3e839a/Userland/Libraries/LibGfx/BMPLoader.cpp#L1493-L1495

[the remark this answers is not in the document]
Comment (nico) [reply Userland/Libraries/LibGfx/BMPLoader.cpp:1498]: …added the newlines.

[the remark this answers is not in the document]
Comment (linusg) [reply Userland/Libraries/LibGfx/BMPLoader.cpp:1498]: thanks, ideally we'd have it consistent everywhere of course but the lack of automated enforcement makes that difficult :smile:

**Review by linusg — APPROVED** [review]

**Pull request merged**